packages/connectors-web/src/alchemySmartWallet.ts

```diff
@@ -1,85 +1,233 @@
+    // A trivial function, but gives a name to indicate how the promise is used.
+    const getSmartWalletClientIfConnected = async (): Promise<
+      SmartWalletClient | undefined
+    > => {
+      return smartWalletClientPromise;
+    };
 
     return {
       id: "alchemySmartWallet",
       name: "Alchemy Smart Wallet",
       type: alchemySmartWallet.type,
-      baseConnector,
+      baseConnector: innerConnector,
 
       async setup(): Promise<void> {
-        await baseConnector.setup?.();
-
-        // Optional function for running when the connector is first created.
-        throw new Error("Not implemented: setup");
+        await innerConnector.setup?.();
+        if (await innerConnector.isAuthorized()) {
+          await initializeSmartWalletClient();
+        }
       },
 
-      async connect(parameters) {
-        await baseConnector.connect(parameters);
-        throw new Error("Not implemented: connect");
+      async connect(params): Promise<{ accounts: Address[]; chainId: number }> {
+        resetState();
+        await innerConnector.connect(params);
+        await initializeSmartWalletClient();
+        const chainId = params?.chainId ?? (await innerConnector.getChainId());
+        const accounts = (await this.getAccounts()) as Address[];
+        config.emitter.emit("connect", { chainId, accounts });
+        return { accounts, chainId };
       },
 
-      async disconnect() {
-        await baseConnector.disconnect?.();
-        throw new Error("Not implemented: disconnect");
+      async disconnect(): Promise<void> {
+        await innerConnector.disconnect?.();
+        resetState();
+        config.emitter.emit("disconnect");
       },
 
-      async getAccounts() {
-        // TODO: Get the account from the signer
-        return [];
+      async getAccounts(): Promise<Address[]> {
+        const smartWalletClient = await getSmartWalletClientIfConnected();
+        const address = smartWalletClient?.account?.address;
+        return address ? [address] : [];
       },
 
-      async getChainId() {
-        // TODO: the connector will need to presist the chain id and instantiate the client/provider with it
-        throw new Error("Not implemented: getChainId");
+      async getChainId(): Promise<number> {
+        return innerConnector.getChainId();
       },
 
-      async getProvider() {
-        throw new Error("Not implemented: getProvider");
+      async getProvider(): Promise<Provider> {
+        const smartWalletClient = await getSmartWalletClientIfConnected();
+        return smartWalletClient?.getProvider() ?? loggedOutProvider;
       },
 
-      async isAuthorized() {
-        throw new Error("Not implemented: isAuthorized");
+      async isAuthorized(): Promise<boolean> {
+        return innerConnector.isAuthorized();
       },
 
-      async switchChain({ chainId }) {
-        void chainId;
-        throw new Error("Not implemented: switchChain");
+      async switchChain(params): Promise<Chain> {
```

Comment:
Haven't had a chance to debug it but I noticed in the demo that when I switch chains when connected with the Alchemy Smart Wallet that this function is called but the new chain ID isn't reflected in the UI

Reply:
Oh I know what happened: it's the isolated event emitter. Previously, the inner connector was publishing its chain change events to the global state, which is wrong but kinda gave the behavior we wanted. Let me fix this.

Reply:
Ohh yeah that makes sense.  Nice find

Reply:
+1